### PR TITLE
Apply gradient transforms after user transforms

### PR DESCRIPTION
**Pull request by andrijapau**
**Context:**  

Currently, user-defined transforms are applied after gradient preprocessing. This ordering is unintuitive and introduces several issues,
- tapes may contain unsupported operations for gradient transforms
- user transforms cannot act on initial higher-order tape structures, as these are decomposed away during preprocessing

Consider this circuit,
```python
import pennylane as qml

dev = qml.device("default.qubit")

@qml.transforms.merge_rotations
@qml.transforms.cancel_inverses
@qml.qnode(dev, diff_method="parameter-shift")
def decorated_circuit(weights):
    qml.RX(weights[0], wires=0)
    qml.RX(weights[1], wires=0)
    qml.RX(weights[2], wires=0)
    return qml.probs(wires=0)

weights = qml.numpy.array([0.1, 0.2, 0.3], requires_grad=True)
decorated_circuit(weights)
```
Before this fix, the transform program in `execute` was,
```
TransformProgram(_expand_transform_param_shift, cancel_inverses, merge_rotations)
```

**Description of the Change:**  

This change proposes applying gradient preprocessing *after* user transforms. 

This will ensure that user transforms can work on the initial tape structure without decomposition interference and that gradient transforms only receive preprocessed tapes with supported operations.

This results in the *correct* transform program,
```
TransformProgram(cancel_inverses, merge_rotations, _expand_transform_param_shift)
```

**Benefits:**  

- **Logical Transform Flow**: Ensures user transforms work as expected on initial structures, like embeddings or entangling layers.
- **Operation Compatibility**: Guarantees that gradient transforms only process compatible operations.
- **Improved User Experience**: Aligns transform order with user expectations and avoids confusion.

**Possible Drawbacks:**  

- **Breaking Change**: This change could disrupt existing setups. We’ll need to:
  - Issue a breaking change alert and be prepared to revert if needed.
  - Coordinate across teams to validate against potential issues with Lightning and Catalyst.
  - Update affected functions (`construct_batch`, `get_transform_program`) and resolve any failing tests or cross-project dependencies.

[sc-72075]

### CODE REVIEW

**Review by albi3ro — APPROVED** [review]
I am quite happy this change worked out as well as it did :)

I'm been thinking a little bit about if we need to test this, and how.  I think we should actually be good to leave this untested. 

What do we actually need to promise to users? That gradient transforms will work (checked) and user transforms get applied (checked).  So I think we should be good to go :)
